### PR TITLE
Add ESLint configuration and use it to replace JSCS / JSHint

### PR DESCRIPTION
**Pull request by robertknight**
This adds a basic ESLint configuration focused on catching errors rather than stylistic issues and replaces the current JSHint / JSCS configurations. Additionally ESLint failures are now errors and will cause Travis builds to fail.

The reason for going with ESLint is because:
 1. It is smarter than JSHint (eg. it recognizes that using a variable requires reading it, not just assigning to it).
 2. The JSCS project recently merged with ESLint (see http://eslint.org/blog/2016/04/welcoming-jscs-to-eslint )
 3. It is very well maintained and documented and consequently the de-facto choice for linting and code style checking in new JS projects.

In addition to the configuration, this PR includes a relatively small set of fixes for basic issues found by the new configuration across the client codebase:
 - Several missing `require('angular')` imports in tests
 - Use for `Array.map` when `Array.forEach` was intended (found by a check that callbacks to array filter/map/reduce functions always return values)
 - A handful of missing `"use strict";` declarations
 - A handful of missing semi-colons

There are other lint rules worth enabling but that can be left for a separate follow-up PR.

### CODE REVIEW

**Comment by nickstenning** [comment]
Looks pretty good. I will say that as a rule I'm not in favour of having lint checks fail the build. Why? Because it actually makes you less likely to enable stricter linting rules, because doing so now require you fix the entire codebase (which carries risks).

I'm ok with the changes made here, but I think it's worth considering whether failing the build is really necessary, or whether you'd get most if not all of the benefit (with, additionally, the ability to easily make the linter stricter in the future) by enabling Hound for this repository, and thus only penalising lint failures in new or changed code.

**Comment by robertknight** [comment]
> Because it actually makes you less likely to enable stricter linting rules,
> because doing so now require you fix the entire codebase (which carries risks).

I strongly feel that enforcing these checks and getting to a clean state, which hasn't been very difficult with ESLint's available rules, is the right thing to do. The problem with warnings and linters is that it adds a secondary tier of "sort-of error" which is easy to ignore and for new contributors it creates confusion about how such issues are supposed to be treated.

If a new awesome linter rule comes along in future that finds a lot of issues that are risky to fix or would add unnecessary noise to the Git history, we can incrementally enable it for the codebase if necessary.

**Comment by nickstenning** [comment]
> I strongly feel that enforcing these checks and getting to a clean state, which hasn't been very difficult with ESLint's available rules, is the right thing to do.

Perhaps, but you can do that without making lint checks fail the build.

At the moment (as is clear from open PRs against this repository) you have to fix all issues before turning on a new lint rule, thus coupling two actions together that don't need to be coupled together.

If you want to do some bulk cleanup of linter issues, that's ok, but you can still do that with a linter like Hound that reports line-by-line lint errors without failing the build. Moreover, you can have distinct PRs that introduce and motivate the linter configuration change, and then fix outstanding issues, perhaps even in chunks small enough to review safely.

Failing the build for linter checks has caused me and projects I've worked on quite a bit of pain in the past, particularly in situations when additional annoyance is the last thing you need (e.g. emergency/hotfix deployments).

Remember: linters are not test suites, and they are sometimes (if rarely) incorrect. Let's keep build status and lint feedback separate.

**Comment by robertknight** [comment]
> At the moment (as is clear from open PRs against this repository) you have to fix all issues
> before turning on a new lint rule, thus coupling two actions together that don't need to be coupled together.

That's true at the moment, but it doesn't have to be the case for new rules. Lint defaults can be overridden at a per-line, per-file and per-directory level.

> Failing the build for linter checks has caused me and projects I've worked on quite a bit of pain
> in the past, particularly in situations when additional annoyance is the last thing you need

I don't think it needs to be painful, provided that lint issues are easy to fix or suppress (they are), the lint runs fast (it does) and you get fast feedback on any issues (ie. you don't find out after a 5 minute build that a lint check failed).

> Remember: linters are not test suites, and they are sometimes (if rarely) incorrect. 

I've been thinking a little more about this. There are three categories of check here:

1. Those which are very likely to indicate an error (such as use of an undefined identifier or unused variable, missing return value in an `Array.filter` callback)
2. Code maintenance recommendations (eg. this function is too complex)
3. Stylistic issues (the indentation in this function is wrong, this function uses var but const/let is preferred).

eg. In a Go codebase 1) Is stuff that would fail with `go build`, 2) are things that `go vet` might complain about and 3) would be deviations from `go fmt` output.

We don't necessarily have to treat these all the same. eg. I quite agree that if we introduce a "prefer const/let to var" rule in future, it wouldn't make sense to require a whole file to be changed before any code changes were made to that file.

> Let's keep build status and lint feedback separate.

I find Hound's style of reporting quite annoying compared to the output of `make lint` here. ESLint's output is also more helpful in that it tells you if it can fix the problem automatically (with `--fix`) and the name of the rule that triggered the issue so a suppression is easy to insert.

So how about this to move forwards:

1) Remove the `npm run lint` check from the test task
2) Add a Travis task which runs the linter and reports output.

**Comment by nickstenning** [comment]
> So how about this to move forwards:
>
> 1) Remove the npm run lint check from the test task
> 2) Add a Travis task which runs the linter and reports output.

I will keep my mouth shut and accede for now ;)

**Comment by robertknight** [comment]
> 1) Remove the npm run lint check from the test task
> 2) Add a Travis task which runs the linter and reports output.

This is now done, see https://travis-ci.org/hypothesis/client/builds/144425027
The `.travis.yml` changes were adapted from something I'd seen another project do. I don't know if there is a slightly shorter way to achieve the same thing.

File: .travis.yml [hunk 1]
```diff
@@ -7,3 +7,13 @@ notifications:
   slack:
     rooms:
       - secure: "iYTftqYShXlMgoeuC7nfmBNwS59ZubIppTD4rpKg6m2PNLal9MtmH5vNlQMfFe+kuLV1pMuPOMfp9y+o3lJGO08lKHKlu+ziAuQcnPW2koR5ViYQcU3NCCZ2Pk4lsolcxwZrS6mbSbL3P/Sv6uYnU+QUNSVPCCeOLxKG+f3Vav+qQx3gduTOLRJEGVtpUpTLmSf8+WZpSM88Y00XK3LVUqOOy2Fq8e13PSERhfWbC391EnuOHpanHtzAdwE5rrWG+EaCYctT//zc/gmy4Cf+coq60RwC0WMhK+zEyLciXRGpBDn0JrGR+24rnI84y9AxdMHXv68OOVx1hZixaKt/8gvpJ78iDRGhEDOrjpGVRrTIApabx02YUsMfJB6XpyCVNRt31LYTW6I4F2+B76UKIEahzdir4LjGbdDhEn+f5adO9bOi46TuCAIhOai220mq/CMv70I4bk8WACuqW1nNq4UaLOZztF0ahmFmHbeGhko5Vlmd3IXhE2tgVN4bk+niOqFWXyJ7Yw/jpe18SkiDoM8jSy6tfQ2ycfwQRHCUHATALAGPYkXFkzKdqrkoRuQBb4hS8MWcOgVdNQwsekul7yR/RFZMBBEj4tUQdBQDP+S1WyCPj9yG5lJ1Ph9jxZ76lBDt3fW2J60UfFtELk4f0f3WR5ipyFSbVqSEmaXad2E="
+script: |
```
Comment (nickstenning) [inline .travis.yml]: Yep, in this particular scenario it's probably easier to do this the way we do it for `h`:

```yml
matrix:
  include:
    - env: ACTION=lint
      script: npm run lint
    - env: ACTION=test
```

Comment (robertknight) [inline .travis.yml]: Hmm, this works but there is an extra unnecessary build in the matrix which has no environment variables set - https://travis-ci.org/hypothesis/client/builds/144432656

Comment (robertknight) [inline .travis.yml]: Looks like https://github.com/travis-ci/travis-ci/issues/4681

Comment (nickstenning) [inline .travis.yml]: Looks like moving the `node_js` into each of the `include:` tasks will sort it.

**Comment by nickstenning** [comment]
LGTM.

**Pull request merged**